common/flatpak-prune.c

```diff
@@ -0,0 +1,826 @@
+  (*name)[32] = (guint8) objtype;
+}
+
+static gint
+flatpak_ostree_name_compare (const FlatpakOstreeObjectName *name_a,
+                             const FlatpakOstreeObjectName *name_b)
+{
+  return memcmp (name_a, name_b, sizeof (FlatpakOstreeObjectName));
+}
+
+static guint
+flatpak_ostree_object_name_hash (gconstpointer a)
+{
+  const FlatpakOstreeObjectName *name = a;
+  const guint8 *data = &(*name)[0];
+
+  /* The checksum is essentially all random, so any 4 bytes of it should
+     be a good hash value. However, we avoid using the first ones, because
+     those are the ones that will be first compared on a hash collision,
+     so if they were always the same that would waste 4 comparisons. */
+  return
+    data[32] |
+    data[31] << 8 |
+    data[30] << 16 |
+    data[30] << 24;
+}
+
+static gboolean
+flatpak_ostree_object_name_equal (gconstpointer a,
+                                  gconstpointer b)
+{
+  const FlatpakOstreeObjectName *name_a = a;
+  const FlatpakOstreeObjectName *name_b = b;
+
+  return flatpak_ostree_name_compare (name_a, name_b) == 0;
+}
+
+/* This is a container for allocating FlatpakOstreeObjectNames in chunks without relocations so
+   that the resulting pointers are stable and can be stored in e.g. a hashtable. */
+
+#define BAG_CHUNK_SIZE 1985 /* nr of objects per chunk in bag, makes chunk fit in 64k with some spare for overhead */
+typedef struct {
+  FlatpakOstreeObjectName *current_chunk; /* Null if non started */
+  gsize current_chunk_used; /* number of used objects in current chunk */
+  GSList *chunks; /* List of allocated chunks */
+  GHashTable *hash;
+} FlatpakOstreeObjectNameBag;
+
+static FlatpakOstreeObjectNameBag *
+object_name_bag_new (void)
+{
+  FlatpakOstreeObjectNameBag *bag = g_new0 (FlatpakOstreeObjectNameBag, 1);
+
+  bag->hash = g_hash_table_new_full (flatpak_ostree_object_name_hash, flatpak_ostree_object_name_equal,
+                                     NULL, NULL);
+
+  return bag;
+}
+
+static void
+object_name_bag_free (FlatpakOstreeObjectNameBag *bag)
+{
+  g_hash_table_unref (bag->hash);
+  g_slist_free_full (bag->chunks, g_free);
+  g_free (bag);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (FlatpakOstreeObjectNameBag, object_name_bag_free)
+
+static gboolean
+object_name_bag_contains (FlatpakOstreeObjectNameBag *bag,
+                          const FlatpakOstreeObjectName *name)
+{
+  return g_hash_table_contains (bag->hash, name);
+}
+
+static void
+object_name_bag_insert (FlatpakOstreeObjectNameBag *bag,
+                        const FlatpakOstreeObjectName *name)
+{
+  FlatpakOstreeObjectName *res;
+
+  if (g_hash_table_contains (bag->hash, name))
+    return;
+
+  if (bag->current_chunk == NULL)
+    {
+      bag->current_chunk = g_new (FlatpakOstreeObjectName, BAG_CHUNK_SIZE);
+      bag->current_chunk_used = 0;
+      bag->chunks = g_slist_prepend (bag->chunks, bag->current_chunk);
+    }
+
+  res = &bag->current_chunk[bag->current_chunk_used++];
```

Comment:
Is there any chance it would make sense to use something like `GINT_TO_POINTER()`, either with 64-bit pointers or with putting the different object types in different hash tables to avoid having to store the 33rd byte? That way we wouldn't have to store the object name and the pointer to the object name separately, unless I'm missing something.

Reply:
Not sure what you mean, the data we need to keep is the sha256, which is 32 bytes, plus a small enum (say 1 byte). A 64bit pointer can only store 8 bytes.

Reply:
My mistake, never mind